File: frontend/app/models/aiprovidermodel/providerset_types.ts
import { ChatCompletionRequestMessage, CompletionRequest, CompletionResponse } from './chat_types';
import { ModelName, ProviderInfo, ProviderName } from './provider_types';

export interface IProviderSetAPI {
	getDefaultProvider(): Promise<ProviderName>;
	setDefaultProvider(provider: ProviderName): Promise<void>;
	getConfigurationInfo(): Promise<Record<string, any>>;
	getProviderInfo(provider: ProviderName): Promise<ProviderInfo>;
	setAttribute(
		provider: ProviderName,
		apiKey?: string,
		defaultModel?: ModelName,
		defaultTemperature?: number,
		defaultOrigin?: string
	): Promise<void>;

	getCompletionRequest(
		provider: ProviderName,
		prompt: string,
		prevMessages?: Array<ChatCompletionRequestMessage>,
		inputParams?: { [key: string]: any },
		stream?: boolean
	): Promise<CompletionRequest>;
	completion(
		provider: ProviderName,
		input: CompletionRequest,
		onStreamData?: (data: string) => Promise<void>
	): Promise<CompletionResponse | undefined>;
}

File: pkg/aiprovider/spec/providerset_types.go
package spec

import "context"

type ProviderSetAPI interface {
	GetDefaultProvider(ctx context.Context) (ProviderName, error)
	SetDefaultProvider(ctx context.Context, provider ProviderName) error
	GetConfigurationInfo(ctx context.Context) (map[string]interface{}, error)
	GetProviderInfo(ctx context.Context, provider ProviderName) (ProviderInfo, error)
	SetAttribute(ctx context.Context, provider ProviderName, apiKey *string, defaultModel *ModelName, defaultTemperature *float64, defaultOrigin *string) error

	GetCompletionRequest(ctx context.Context, provider ProviderName, prompt string, prevMessages []ChatCompletionRequestMessage, inputParams map[string]interface{}, stream bool) (CompletionRequest, error)
	Completion(ctx context.Context, provider ProviderName, input CompletionRequest, onStreamData func(data string) error) (*CompletionResponse, error)
}
